bftengine/tests/controllerWithSimpleHistory/ControllerWithSimpleHistory_test.cpp

```diff
@@ -145,7 +147,319 @@ TEST(ControllerWithSimpleHistory, onNewView_not_primary) {
+ public:
+  string toString() const override { return "IShareSecretKeyDummy"; }
+};
+
+class IShareVerificationKeyDummy : public IShareVerificationKey {
+ public:
+  string toString() const override { return "IShareVerificationKeyDummy"; }
+};
+
+class ThreshSigMock : public IThresholdSigner {
+ public:
+  IShareSecretKeyDummy is;
+  IShareVerificationKeyDummy isv;
+  virtual int requiredLengthForSignedData() const { return 5; };
+  virtual void signData(const char *hash, int hashLen, char *outSig, int outSigLen){};
+
+  virtual const IShareSecretKey &getShareSecretKey() { return is; };
+  virtual const IShareVerificationKey &getShareVerificationKey() const { return isv; };
+  const std::string getVersion() const { return "v"; }
+  void serializeDataMembers(std::ostream &) const {}
+  void deserializeDataMembers(std::istream &) {}
+  const IShareSecretKey &getShareSecretKey() const { return is; }
+};
+
+// Test - Downgrade from OPTIMISTIC_FAST to FAST_WITH_THRESHOLD then upgrade to OPTIMISTIC_FAST
+// E.L Logic: Not Possible, Bug ?
+
+// Test - Downgrade from OPTIMISTIC_FAST to SLOW then OPTIMISTIC_FAST
+// Logic:
+// 1)Construct controller with C == 0 i.e. disable FAST_WITH_THRESHOLD path.
+// 2)Run loop for EvaluationPeriod times:
+//  - call to onStartingSlowCommit for more than factor * EvaluationPeriod.
+// 3)Run loop for EvaluationPeriod times:
+//  - call onMessage for each replica to satisfy full cooporation.
+TEST(ControllerWithSimpleHistory, upgrade_slow_to_optimistic) {
+  uint16_t C = 0;
+  uint16_t F = 1;
+  ReplicaId replicaId = 0;
+  ViewNum initialView = 0;
+  SeqNum initialSeq = 0;
+  ControllerWithSimpleHistory cwsh{C, F, replicaId, initialView, initialSeq};
+
+  ASSERT_EQ(CommitPath::OPTIMISTIC_FAST, cwsh.getCurrentFirstPath());
+
+  // Factor + 1, in order to trigger the degradation.
+  auto slowStartCount =
+      ControllerWithSimpleHistory_debugDowngradeFactor * ControllerWithSimpleHistory::EvaluationPeriod + 1;
+  bool changed{false};
+  auto prePrepareTime = std::chrono::steady_clock::now() - std::chrono::milliseconds(10);
+  for (auto i = (size_t)1; i <= ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    if (slowStartCount-- > 0) cwsh.onStartingSlowCommit((SeqNum)i);
+  }
+
+  ASSERT_EQ(true, changed);
+  ASSERT_EQ(CommitPath::SLOW, cwsh.getCurrentFirstPath());
+
+  char buf[5] = {'m', 'o', 's', 'h', 'e'};
+  Digest d{buf, 5};
+  ThreshSigMock th;
+  for (auto i = (size_t)65; i < 65 + ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    for (auto id : {1, 2, 3}) {
+      auto p = impl::PreparePartialMsg::create(0, i, id, d, &th);
+      cwsh.onMessage(p);
+      delete p;
+    }
```

Comment:
so here the upgrade happens because we don't call `cwsh.onNewSeqNumberExecution` for a certain number of times, right?

Reply:
Actually it's because we call `cwsh.onMessage(p)` per replica, which eventually satisfies the full cooperation  requirement.